modules/tinymce/src/core/main/ts/delete/TableDelete.ts

```diff
@@ -242,68 +240,70 @@ const isDeleteOfLastCharPos = (fromCaption: SugarElement<HTMLTableCaptionElement
   fromCaption: SugarElement<HTMLTableCaptionElement>,
   from: CaretPosition
-): Optional<boolean> =>
+): Optional<() => void> =>
   CaretFinder.navigate(forward, editor.getBody(), from).bind(
     (to) => isDeleteOfLastCharPos(fromCaption, forward, from, to) ?
       emptyCaretCaption(editor, fromCaption) :
       validateCaretCaption(rootElm, fromCaption, to)
-  ).or(Optional.some(true));
+  ).getOr(Optional.some(Fun.noop));
 
-const deleteCaretCells = (editor: Editor, forward: boolean, rootElm: SugarElement<Node>, startElm: SugarElement<Node>): boolean => {
+const deleteCaretCells = (editor: Editor, forward: boolean, rootElm: SugarElement<Node>, startElm: SugarElement<Node>): Optional<() => void> => {
   const from = CaretPosition.fromRangeStart(editor.selection.getRng());
   return getParentCell(rootElm, startElm).bind(
     (fromCell) => Empty.isEmpty(fromCell) ?
       emptyElement(editor, fromCell) :
       deleteBetweenCells(editor, rootElm, forward, fromCell, from)
-  ).getOr(false);
+  );
 };
 
 const deleteCaretCaption = (
   editor: Editor,
   forward: boolean,
   rootElm: SugarElement<Node>,
   fromCaption: SugarElement<HTMLTableCaptionElement>
-): Optional<boolean> => {
+): Optional<() => void> => {
   const from = CaretPosition.fromRangeStart(editor.selection.getRng());
   return Empty.isEmpty(fromCaption) ?
     emptyElement(editor, fromCaption) :
     deleteCaretInsideCaption(editor, rootElm, forward, fromCaption, from);
 };
 
-const isNearTable = (forward: boolean, pos: CaretPosition): boolean =>
-  forward ? isBeforeTable(pos) : isAfterTable(pos);
+const isNearTable = (forward: boolean, pos: CaretPosition): Optional<() => void> => {
+  const isNear = forward ? isBeforeTable(pos) : isAfterTable(pos);
+  return isNear ? Optional.some(Fun.noop) : Optional.none();
+};
 
-const isBeforeOrAfterTable = (editor: Editor, forward: boolean): boolean => {
+const isBeforeOrAfterTable = (editor: Editor, forward: boolean): Optional<() => void> => {
   const fromPos = CaretPosition.fromRangeStart(editor.selection.getRng());
 
-  return isNearTable(forward, fromPos) ||
+  return isNearTable(forward, fromPos).orThunk(() =>
     CaretFinder.fromPosition(forward, editor.getBody(), fromPos)
-      .exists((pos) => isNearTable(forward, pos));
+      .bind((pos) => isNearTable(forward, pos)));
 };
 
-const deleteCaret = (editor: Editor, forward: boolean, startElm: SugarElement<Node>): boolean => {
+const deleteCaret = (editor: Editor, forward: boolean, startElm: SugarElement<Node>): Optional<() => void> => {
   const rootElm = SugarElement.fromDom(editor.getBody());
 
   return getParentCaption(rootElm, startElm).fold(
-    () => deleteCaretCells(editor, forward, rootElm, startElm) || isBeforeOrAfterTable(editor, forward),
-    (fromCaption) => deleteCaretCaption(editor, forward, rootElm, fromCaption).getOr(false)
+    () => deleteCaretCells(editor, forward, rootElm, startElm).orThunk(() => isBeforeOrAfterTable(editor, forward)),
```

Comment:
In this file I think this is where you need to thunk not on all predicates.

Reply:
I'm not sure what you mean?

Reply:
`isBeforeOrAfterTable` is a predicate and should remain a predicate this is not using options as boolean but rather producing a thunk when something needs to be done or not. That thunk can either be just block the browser native action a Fun.noop or some action that actually alters the dom.

Reply:
See if you like that better.